Run unittests on Python 3.12

mweinelt opened the pull request:
- Test pytorch stable and nightly
- Use uv's pip interface for faster dependency resolution
- Migrate away from deprecated `py.test` executable
- Convert various options to long-opts, makes them easier to understand

mweinelt committed:
Enable test matrix for Python 3.9 to 3.13

- Test pytorch stable and nightly
- Use uv's pip interface for faster dependency resolution
- Migrate away from deprecated `py.test` executable
- Convert various options to long-opts, makes them easier to understand

mweinelt committed:
Replace setup.py with pyproject.toml (PEP621)

This is required for using uv I'm afraid, but the changes should increase
the QoL of the life, since we can rely on a no-code approach to define
the package.

This is based on https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html

mpariente commented on .github/workflows/test_package.yml:
```diff
@@ -9,16 +9,18 @@ jobs:
       matrix:
-        python-version: [3.6]  #, 3.7, 3.8]
-        pytorch-version: ["1.8.0", "nightly"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
```
One version of Python only to save some compute.

mweinelt replied:
Okay, trying to figure out what versions actually work. Then I'll reduce the scope to one of those.

mweinelt replied:
3.12 seems to be the latest version that works, so I chose that.

The full test matrix takes 1m20s fwiw, which is probably fine.

mpariente commented on .github/workflows/test_package.yml:
```diff
@@ -9,16 +9,18 @@ jobs:
-        python-version: [3.6]  #, 3.7, 3.8]
-        pytorch-version: ["1.8.0", "nightly"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
+        pytorch-version: ["stable", "nighly"]
```
Let's keep a fixed version of PyTorch (instead of stable) besides nightly

mweinelt replied:
1.8.0, because it is the oldest supported one?
2.5.1 because it is the latest stable release?

mpariente replied:
2.5.1 and nighly will be perfect.

mpariente commented on pyproject.toml:
```diff
@@ -1,9 +1,47 @@
+[project]
+name = "asteroid-filterbanks"
+authors = [
+    {name = "Manuel Pariente", email="manuel.pariente@loria.fr"}
```
Change to pariente.mnl@gmail.com please.

mweinelt replied:
Done

mpariente reviewed (commented):
LGTM in general, thanks ! 

What's the advantage of UV, please ?

mweinelt commented:
> What's the advantage of UV, please ?

Pip compatible interface but [10-100x faster](https://github.com/astral-sh/uv/blob/main/BENCHMARKS.md), thus saving on compute and giving faster response times.

mweinelt commented:
Okay. Looks like tests to 2.5.1 are running on 3.9-3.12. The nightly torch tests fail during dependency resolution, because for some reason they want to include `llvmlite==0.36.0` (via librosa, which has no upper bound), which does not support Python 3.10+.

Time for dinner.

mweinelt committed:
requirements.txt: Require newer numba version for 3.10+

mweinelt committed:
Run tests only on Python 3.12

Running on just one version was requested, as to reduce the compute we
consume through the CI.

mweinelt committed:
Run test_torch_sft on Python 3.12

mweinelt commented:
I had to pin to a newer numba version, to get a llvmlite version compatible with 3.10 and newer.

Now all versions 3.9-3.12 are working with torch 2.5.1 and nightly.

The `test_torch_stft` test however seems broken with torch==2.5.1.
